### PR TITLE
#168781689- admin can assign role to user

### PR DESCRIPTION
**Pull request by sabin18**
#### What does this PR do?
PR to add role settings and admin controller.
#### Description of Task to be completed?
- created get all users function. 
- created  get one user function.
- created   get roles function.
- created  assign roles function.
- created  roles seeders.
- created  admin permission.
#### How should this be manually tested?
- clone the repository : git clone https://github.com/andela/caret-bn-backend.git
- checkout to branch ft-User-Role-Settings-168781689
- run sequelize db:migrate
- run sequelize db:seed:all
- Login as admin
- Take the token to your headers in authorization as Bearer Token to each routes of this feature.
- Get all users : /api/v1/admin/users/
- Get all roles : /api/v1/admin/roles/
- Get one user or single user : /api/v1/admin/users/:id   :id=>user id
- Assign role to user: /api/v1/admin/roles/assign/:id                 :id=>user id
- to add role to user you write the name of role you want to give to the user in the body as Role.

#### Any background context you want to provide?
`N/A`
#### What are the relevant pivotal tracker stories?
[168781689](https://www.pivotaltracker.com/story/show/168781689)
#### Screenshots (if appropriate)
Screen Shot 2019-10-17 at 22 35 28
Screen Shot 2019-10-17 at 22 38 31
Screen Shot 2019-10-22 at 17 04 51
Screen Shot 2019-10-22 at 17 20 03

#### Questions:
`N/A`

### CODE REVIEW

**Review by alainmateso — CHANGES_REQUESTED** [review]
Good Job Sabin! I would like for you to impolement some changes though. Like re-using the existing token validation that handles even incorrect tokens(which yours doen't) instead of creating a new one as well as implementing your input validations into the existing validation class.

**Review by alainmateso — APPROVED** [review]
Great job Sabin

**Review by nignanthomas — CHANGES_REQUESTED** [review]
Woow, great job @sabin18 !
However
- remove the **password** in the returned users and you shall be good.
Even a Super Admin can't see users' passwords.
- the endpoints break when **id !== integer** eg: /admin/users/wa
I have a checkId middleware that you will be able to reuse when my feature is merged

**Review by nignanthomas — APPROVED** [review]

File: src/controllers/adminController.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  static GetRoles(req, res) {
+    models.Role.findAll().then(roles => responseUtil(res, 200, strings.users.success.FOUND, roles));
+  }
+
+  // assign roles to user
+  static AssignRoles(req, res) {
+    const { Role } = req.body;
+    const { id } = req.params;
+    models.users.findOne({ where: { id }, }).then(user => {
+
```
Comment (nakiwuge) [inline src/controllers/adminController.js:40]: Remove this line

File: src/controllers/adminController.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      if (user == null) {
```
Comment (nakiwuge) [inline src/controllers/adminController.js]: You can write if `!user` instead if `user == null`

File: src/controllers/adminController.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        return responseUtil(res, 404, strings.users.error.USER_NOT_FOUND);
+      }
+
```
Comment (nakiwuge) [inline src/controllers/adminController.js]: Remove this line

File: src/controllers/adminController.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      models.Role.findOne({ where: { name: Role }, }).then(role => {
+
```
Comment (nakiwuge) [inline src/controllers/adminController.js]: Remove this line

File: src/database/migrations/20191007121953-addPasswordAndVerificationColumns.js [hunk 5]
```diff
@@ -11,9 +11,9 @@ module.exports = {
       defaultValue: false
     }, { transaction: t }),
     queryInterface.addColumn('users', 'role', {
-      type: Sequelize.STRING,
+      type: Sequelize.INTEGER,
```
Comment (nakiwuge) [inline src/database/migrations/20191007121953-addPasswordAndVerificationColumns.js:14]: This should be a foreign key

File: src/middlewares/permissions.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import jwt from 'jsonwebtoken';
+import dotenv from 'dotenv';
+import models from '../database/models';
+import strings from '../utils/stringsUtil';
+import responseUtil from '../utils/responseUtil';
+
+dotenv.config();
+
```
Comment (nakiwuge) [inline src/middlewares/permissions.js]: Remove this line

Comment (nakiwuge) [inline src/middlewares/permissions.js]: Remove this line

File: src/routes/api/admin.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import permissions from '../../middlewares/permissions';
+import InputValidation from '../../middlewares/inputValidation';
+import validateToken from '../../middlewares/auth/validateToken';
+import checkId from '../../middlewares/checkId';
+
+const {
+  GetAllUsers, GetOneUsers, GetRoles, AssignRoles
+} = AdminController;
+const { AdminAccess } = permissions;
+
```
Comment (nakiwuge) [inline src/routes/api/admin.js]: Remove this line

File: src/routes/api/admin.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+const { validateRole } = InputValidation;
+
```
Comment (nakiwuge) [inline src/routes/api/admin.js]: Remove this line

File: src/routes/api/admin.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+const router = express.Router();
+
+router.get('/users/', validateToken, AdminAccess, GetAllUsers);
+router.get('/users/:id', validateToken, AdminAccess, checkId, GetOneUsers);
+router.get('/roles/', validateToken, AdminAccess, GetRoles);
+router.patch('/assign/:id', validateToken, AdminAccess, checkId, validateRole, AssignRoles);
```
Comment (nakiwuge) [inline src/routes/api/admin.js]: change this route to `roles/assign/:id`

File: src/tests/mockdata.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
+const testdata = {
+  validSignup: {
+    username: 'caretdevs',
+    email: 'teamcaret@gmail.com',
+    password: 'Pa$5W0rd',
+    confirmPassword: 'Pa$5W0rd',
+  },
+
+  verifyUser: { id: 3, isVerified: false },
+
```
Comment (nakiwuge) [inline src/tests/mockdata.js:10]: Remove this line

File: src/tests/mockdata.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+  validSignup: {
+    username: 'caretdevs',
+    email: 'teamcaret@gmail.com',
+    password: 'Pa$5W0rd',
+    confirmPassword: 'Pa$5W0rd',
+  },
+
+  verifyUser: { id: 3, isVerified: false },
+ 
+
```
Comment (nakiwuge) [inline src/tests/mockdata.js:11]: Remove this line

**Review by nakiwuge — CHANGES_REQUESTED** [review]
Good work @sabin18. Kindly work on the feedback below.
- Add an association between the roles and users table
- Return the roles object as the value to the key `role` when getting a user
I have left other comments on the pr.

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**